File: demo/backend/middlewares/session.js
const jwt = require("jsonwebtoken");
const { ErrorCode } = require("../constants/errorCodes");

const SESSION_SECRET = process.env.SESSION_SECRET || "default-secret-key";

const validateSession = (req, res, next) => {
  const authorization = req.headers.authorization;

  if (!authorization || !authorization.startsWith("Bearer ")) {
    return res.status(400)
    .json({
      success: false,
      error: {
        code: ErrorCode.FIELD_CANNOT_BE_BLANK,
        message: "Authorization is missing or invalid",
        message_title: "Unauthorized",
      },
    });
  }

  const sessionToken = authorization.split(" ")[1];

  try {
    const sessionData = jwt.verify(sessionToken, SESSION_SECRET);

    if (sessionData.ip !== req.ip) {
      return res.status(403)
      .json({
        success: false,
        error: {
          code: ErrorCode.INVALID_AUTHORIZATION,
          message: "Your session is invalid or expired. Please reload the page and try again!",
          message_title: "Unauthorized",
        },
      });
    }

    req.session = sessionData;
    next();
  } catch (err) {
    return res.status(403)
    .json({
      success: false,
      error: {
        code: ErrorCode.INVALID_AUTHORIZATION,
        message: "Your session is invalid or expired. Please reload the page and try again!",
        message_title: "Unauthorized",
      },
    });
  }
};

module.exports = {
  validateSession,
};

File: demo/backend/middlewares/session.py
from flask import request, jsonify
import jwt
from constants.error_codes import ErrorCode
from config import Config

def validate_session(func):
    def wrapper(*args, **kwargs):
        auth_header = request.headers.get("Authorization")
        if not auth_header or not auth_header.startswith("Bearer "):
            return jsonify({
                "success": False,
                "error": {
                    "code": ErrorCode.FIELD_CANNOT_BE_BLANK,
                    "message": "Authorization is missing or invalid",
                    "message_title": "Unauthorized",
                }
            }), 400

        session_token = auth_header.split(" ")[1]
        try:
            session_data = jwt.decode(session_token, Config.SESSION_SECRET, algorithms=["HS256"])
            if session_data.get("ip") != request.remote_addr:
                raise jwt.InvalidTokenError
            request.environ["session_data"] = session_data
            return func(*args, **kwargs)
        except jwt.ExpiredSignatureError:
            return jsonify({
                "success": False,
                "error": {
                    "code": ErrorCode.INVALID_AUTHORIZATION,
                    "message": "Session expired",
                    "message_title": "Unauthorized",
                }
            }), 403
        except jwt.InvalidTokenError:
            return jsonify({
                "success": False,
                "error": {
                    "code": ErrorCode.INVALID_AUTHORIZATION,
                    "message": "Invalid session token",
                    "message_title": "Unauthorized",
                }
            }), 403

    return wrapper
